gnocchi/metric/v1/resources/requests.go

```diff
@@ -99,11 +102,19 @@ type CreateOpts struct {
 	// UserID is the Identity user of the resource.
 	UserID string `json:"user_id,omitempty"`
 
-	// StartedAt is a resource creation timestamp.
-	StartedAt string `json:"started_at,omitempty"`
+	// StartedAtTimeStamp is a resource creation timestamp. It's contents will be converted into the string
+	// and then transferred to the private "startedAt" field.
+	StartedAtTimeStamp *time.Time `json:"-"`
+
+	// StartedAtString is a string representation of the "StartedAt" field.
+	StartedAtString string `json:"started_at,omitempty"`
```

Comment:
Let's drop these `String` fields; just have `StartedAt` (no `TimeStamp`). Same with `EndedAt`.

Reply:
@jrperritt @jtopjian I have a problem with that. If I'm providing a time.Time or a &time.Time into Gophercloud's `func BuildRequestBody`, it returns error at the [L156](https://github.com/gophercloud/gophercloud/blob/master/params.go#L156):
```
2018/01/17 22:43:55 json: cannot unmarshal string into Go value of type map[string]interface {}
```
It's printing a string representation of a time.Time at [L152](https://github.com/gophercloud/gophercloud/blob/master/params.go#L152) if it's uncommented:
```
string(b): "2018-01-02T23:23:34Z"
```

Can you give an advice what is better to do?
Or I should create an issue inside Gophercloud's main repo and we will talk about it inside and leave that PR for now?

Reply:
@ozerovandrei Try calling `b, err := gophercloud.BuildRequestBody(opts, "")` at the beginning of the function and then work on all of the special/unique fields.

Reply:
@jtopjian that's the thing. When I'm providing an opts and some filed in the opts has the `time.Time` or the `&time.Time` type, the `BuildRequestBody` throws an error that is shown in my previous comment.

Reply:
@ozerovandrei Oops - I'm really sorry about that. 

You're right: there's a bug in `gophercloud/gophercloud/params.go`. 

I see what you were trying to accomplish with the `StartedAtString` fields now. Instead of using those fields, the following also works:

```go
func (opts CreateOpts) ToResourceCreateMap() (map[string]interface{}, error) {
  var startedAt, endedAt string

  if opts.StartedAt != nil {
    startedAt = opts.StartedAt.Format(gnocchi.RFC3339NanoTimezone)
    opts.StartedAt = nil
  }

  if opts.EndedAt != nil {
    endedAt = opts.EndedAt.Format(gnocchi.RFC3339NanoTimezone)
    opts.EndedAt = nil
  }

  b, err := gophercloud.BuildRequestBody(opts, "")
  if err != nil {
    return nil, err
  }

  if startedAt != "" {
    b["started_at"] = startedAt
  }

  if endedAt != "" {
    b["ended_at"] = endedAt
  }

  if opts.ExtraAttributes != nil {
    for key, value := range opts.ExtraAttributes {
      b[key] = value
    }
  }

  return b, nil
}
```

_but_ that's a hack and I think we should fix `params.go` instead.

@jrperritt The `time.Time` field is getting passed back in `BuildRequestBody` [here](https://github.com/gophercloud/gophercloud/blob/master/params.go#L138). What's the best way to handle this? This feels just as hacky as the above:

```go
if v.Type().String() != "*time.Time" {
  //fmt.Printf("Calling BuildRequestBody with:\n\tv: %+v\n\tf.Name:%s\n", v.Interface(), f.Name)
  _, err := BuildRequestBody(v.Interface(), f.Name)
  if err != nil {
    return nil, err
  }
}
```

Reply:
@jtopjian I think in `BuildRequestBody`, while iterating the struct fields, we can check toward the top of the loop if the `json` struct tag is "-" and `continue` if it is

Reply:
@jrperritt Thank you! 😄 

@ozerovandrei This should now be fixed in https://github.com/gophercloud/gophercloud/pull/730.  The `ToResourceCreateMap` should now be able to work as:

```go
// ToResourceCreateMap constructs a request body from CreateOpts.
func (opts CreateOpts) ToResourceCreateMap() (map[string]interface{}, error) {
  b, err := gophercloud.BuildRequestBody(opts, "")
  if err != nil {
    return nil, err
  }

  if opts.StartedAt != nil {
    b["started_at"] = opts.StartedAt.Format(gnocchi.RFC3339NanoTimezone)
  }

  if opts.EndedAt != nil {
    b["ended_at"] = opts.EndedAt.Format(gnocchi.RFC3339NanoTimezone)
  }

  if opts.ExtraAttributes != nil {
    for key, value := range opts.ExtraAttributes {
      b[key] = value
    }
  }

  return b, nil
}
```

Reply:
@jtopjian thank you!